lib/services/app.go

```diff
@@ -158,3 +169,217 @@ func UnmarshalAppServer(data []byte, opts ...MarshalOption) (types.AppServer, er
+	}
+
+	// If we ended up with more than one app from the service, we'll have port name in the app name to distinguish them.
+	// We can only do this now because we don't know how many apps there will be until all ports are processed by errgroup.
+	for i := range apps {
+		if len(apps) > 1 {
+			apps[i].SetName(apps[i].(*types.AppV3).Spec.PublicAddr)
+		}
+		apps[i].(*types.AppV3).Spec.PublicAddr = "" // Clear temporary used for name field.
+	}
+
+	return apps, nil
+}
+
+// ProtocolChecker is an interface used to check what protocol uri serves
+type ProtocolChecker interface {
+	CheckProtocol(uri string) string
+}
+
+func getServiceFQDN(s v1.Service) string {
+	host := s.GetName()
+	if s.Spec.Type == v1.ServiceTypeExternalName {
+		host = s.Spec.ExternalName
+	}
+	return fmt.Sprintf("%s.%s.svc.cluster.local", host, s.GetNamespace())
+}
+
+const (
+	protoHTTPS = "https"
+	protoHTTP  = "http"
+	protoTCP   = "tcp"
+)
+
+func getAppURI(serviceFQDN string, port v1.ServicePort, protocolAnnotation string, pc ProtocolChecker) string {
+	constructURI := func(p string) string { return fmt.Sprintf("%s://%s:%d", p, serviceFQDN, port.Port) }
+
+	proto := strings.ToLower(protocolAnnotation)
+	if proto == protoHTTP || proto == protoHTTPS {
+		return constructURI(proto)
+	}
+
+	if port.AppProtocol != nil {
+		switch strings.ToLower(*port.AppProtocol) {
+		case protoHTTP:
+			return constructURI(protoHTTP)
+		case protoHTTPS:
+			return constructURI(protoHTTPS)
+		}
+	}
+
+	if port.Port == 443 || strings.ToLower(port.Name) == protoHTTPS {
+		return constructURI(protoHTTPS)
+	}
+
+	if pc != nil {
+		result := pc.CheckProtocol(fmt.Sprintf("%s:%d", serviceFQDN, port.Port))
+		if result != protoTCP {
+			return constructURI(result)
+		}
+	}
+
+	if port.Port == 80 || port.Port == 8080 || strings.ToLower(port.Name) == protoHTTP {
+		return constructURI(protoHTTP)
+	}
+
+	return constructURI(protoTCP)
+}
+
+func getAppRewriteConfig(annotations map[string]string) (*types.Rewrite, error) {
+	rewrite := annotations[types.DiscoveryAppRewriteLabel]
+	if rewrite == "" {
+		return nil, nil
+	}
+
+	rw := types.Rewrite{}
+	reader := strings.NewReader(rewrite)
+	decoder := kyaml.NewYAMLOrJSONDecoder(reader, 32*1024)
+	err := decoder.Decode(&rw)
+	if err != nil {
+		return nil, trace.Wrap(err, "failed decoding rewrite config")
+	}
+
+	return &rw, nil
+}
+
+func getAppName(serviceName, namespace, clusterName, portName string) string {
+	clusterName = strings.ReplaceAll(clusterName, ".", "-")
+	if portName != "" {
+		return fmt.Sprintf("%s-%s-%s-%s", serviceName, portName, namespace, clusterName)
+	}
+	return fmt.Sprintf("%s-%s-%s", serviceName, namespace, clusterName)
+}
+
+func getAppLabels(serviceLabels map[string]string, clusterName string) map[string]string {
+	result := make(map[string]string, len(serviceLabels)+1)
+
+	for k, v := range serviceLabels {
+		if !types.IsValidLabelKey(k) {
+			logrus.Debugf("Skipping label %q as invalid while creating app labels from service", k)
```

Comment:
use the configured logger since it has the preferences of the user

Reply:
I think this `IsValidLabelKey()` is not required, since what teleport allows in label keys is actually a superset of what Kubernetes allows. So I'm inclined to remove it.

Reply:
I return error now instead of just skipping label.